pwdx.1
.\" Man page for pwdx
.\" Licensed under version 2 of the GNU General Public License.
.\" Copyright 2004 Nicholas Miell.
.\" Based on the pmap(1) man page by Albert Cahalan.
.\"
.TH PWDX "1" "June 2011" "procps-ng" "User Commands"
.SH NAME
pwdx \- report current working directory of a process
.SH SYNOPSIS
.B pwdx
[\fIoptions\fR] \fIpid\fR [...]
.SH OPTIONS
.TP
\fB\-V\fR, \fB\-\-version\fR
Output version information and exit.
.TP
\fB\-h\fR, \fB\-\-help\fR
Output help screen and exit.
.SH "SEE ALSO"
.BR ps (1),
.BR pgrep (1)
.SH STANDARDS
No standards apply, but pwdx looks an awful lot like a SunOS command.
.SH AUTHOR
.UR nmiell@gmail.com
Nicholas Miell
.UE
wrote pwdx in 2004.
.SH "REPORTING BUGS"
Please send bug reports to
.UR procps@freelists.org
.UE


NEWS
procps-ng-3.3.3
---------------
  * watch -g command repeats until something changes
  * Changed labels of kB etc to KiB Debian #662786
  * top supports hotplugged memory and cpus Debian #351934, #459287
  * top %cpu not distorted when switch to/from threads Debian #441166
  * top can read old toprc configuration files Debian #651213, #651863
  * top supports input paste once again Debian #663334
  * top -p avoids SEGV when pid doesn't exist Debian 668335

  * Documentation catch up (fixed prior to 3.3.3)
    . top repeating "%" wastes space Debian #322984
    . top -U user name parse error Debian #623200
    . top memory leaks Debian #627257
    . top impossible task swap statistics Debian #628462
    . top forest view segfault Debian #650864

procps-ng-3.3.1 --> procps-ng-3.3.2
--------------------------------

  * Redefined library to use version-info
  * NLS added! Programs can use po files to localise the output to suite
    any supported languages.
  * Imported a bunch of distribution patches:
      watch: support unicode
      watch: add precision wait time option -p
      watch: interpret ANSI color code sequences
      watch: add -exec and -beep flags and has better quoting
      w: use environment to set user and from/host column widths
      w: use COLUMNS environment if TIOCGWINSZ fails
      w: bassman emulation with -o option
      vmstat: do not scale si/so just like bi/bo
      libprocps-ng: sysinfo.c: truncate the vmstat figure to 32 bits
      tload: remote unneeded optarg and optind variables
      sysctl: fix up some option processing
      skill: kill prints perror
      skill: do not treat skill null parameter as 0
      skill: fix too greedy option parser
      libprocps-ng: readproc.c: some type conversion help
      ps: rename SZ to SIZE
      ps: add sorting to %mem for ps
      pmap: provide information for -x option
      pgrep: distinguish between invalid commandline parameters and '-?'
      pgrep: fix compiler warning saved_start_time might be used uninitialized
      pgrep: add -c option for counting number of matched proceesses
      pwdx & libprocps-ng: Hurd does not have MAX_PATH defined
      ps: --sort does not work with time argument
      skill: add CR to warning line
      contrib: minimal ps: define mips PAGE_SIZE
      libproc-ng: prettyfy proc mount messages
      ps: add build option to disable ps option warning
      libproc-ng: support building without WCHAR support
      sysctl: remove index() for buildroot
  * top now has a search capability
  * User import is validated better to stop negative values for intervals etc

procps-ng-3.3.0 --> procps-ng-3.3.1
--------------------------------

  * Added DejaGNU regression testing
  * Fixed pgrep -u not finding processes Debian#649591
  * Fixed pgrep crashing
  * vmstat -p <part> finds partitions. Was Debian patch vmstat_part_format
    fixes closed bugs RH#485243 and Debian#588677
  * watch 8-bit clean, Was Debian patch watch_8bitchar
  * slabtop prints plain ASCII in once mode

procps-3.2.8 --> procps-ng-3.3.0
------------------------------------------------------
Debian, Fedora and openSUSE fork of procps.
https://gitorious.org/procps

environment/build changes:
 . autotools integrated
 . procps-ng specific options
     --enable-oomem  (add out-of-memory flds)
     --disable-kill  (do not build kill pgm)
     --enable-w-from (from as default for w)

library changes:
 . added control groups support
 . added supplementary groups support
 . introduced proc_t memory management
 . enabled hot-plugged cpu support
 . introduced new api's
     readeither   (readproc.c)
     readproctab3 (readproc.c)
     escaped_copy (escape.c)

free changes:
 . rewritten/modernized
     improved help
     includes long options
     added new processing options
 . updated manual page

ps changes:
 . added new fields
     cgroups,etimes,supgid,supgrp,tgid
 . exploited new library api's

top changes:
 . redesigned fields management
 . added new fields
     cgroups,gid,nmin,nth,pgrp,ruid,
     sid,suid,supgids,supgrps,suser,
     tgid,tpgid
 . added horizontal/vertical scrolling
 . added flexible output override via -w
 . added COMMAND column forest view via V
 . added hot-plugged cpu/memory support
 . improved user filtering via u|U
 . improved signal handling
 . restored zero delay functionality
 . introduced true line input editing
 . exploited new library api's
 . improved performance up to 300%

Debian, Fedora and openSUSE fork of procps.
https://gitorious.org/procps
------------------------------------------------------

procps-3.2.8 --> procps-3.2.9 (pending)

top: fix an aliasing problem -- thanks David Owen

procps-3.2.7 --> procps-3.2.8

ps: allow "+" in sort specifications, as in man page     rh208217
ps: recognize SCHED_ISO and SCHED_IDLE
ps: document SCHED_BATCH and add a "see also" for stime
ps: man page less ambiguous
top: normal exit code should be 0     #341272 #354255 rh199174 suse493210
top: misc fixes
pgrep: usage error should exit with 2       #413383
vmstat: use EXIT_FAILURE -- thanks Yoshio Nakamura   #425492
sysctl: fix crash -- thanks Steinar Gunderson      #423704
watch: tolerate umlauts                          #207103
pmap: range limits with -A low,high
update /dev/tty* info to May 2009 devices.txt
don't read off end of string const                rh469495 rh498182

procps-3.2.6 --> procps-3.2.7

top: document H option -- thanks Tony Ernst
top: terabytes -- thanks Tony Ernst
top: CPU usage column width fixes -- thanks Tony Ernst
top: *roff change                                   #326517
ps: SCHED_BATCH is B
ps: fix s format (signals) output with thread display
watch: avoid integer overflow for the time delay
pwdx: buffer overflow fixed -- thanks Ulf Harnhammar
procps.spec needed a slash -- thanks Jesse Brandeburg
w: stale utmp entries snuck in via uninitialized var -- thanks Robert A Basch
pgrep/pkill: fix some realloc-related crashes       #353894

pwdx.c
/*
 * pwdx.c - print process working directory
 * Copyright 2004 Nicholas Miell
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#include "proc/version.h"
#include "c.h"
#include "nls.h"
#include "xalloc.h"
#include "fileutils.h"

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
	fputs(USAGE_HEADER, out);
	fprintf(out, _(" %s [options] pid...\n"), program_invocation_short_name);
	fputs(USAGE_OPTIONS, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("pwdx(1)"));

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int check_pid_argument(char *input)
{
	int skip = 0;
	long pid;
	char *end = NULL;

	if (!strncmp("/proc/", input, 6))
		skip = 6;
	pid = strtol(input + skip, &end, 10);

	if (errno || input + skip == end || (end && *end))
		return 1;
	if (pid < 1)
		return 1;
	return 0;
}

int main(int argc, char *argv[])
{
	char ch;
	int retval = 0, i;
	int alloclen = 128;
	char *pathbuf;

	static const struct option longopts[] = {
		{"version", no_argument, 0, 'V'},
		{"help", no_argument, 0, 'h'},
		{NULL, 0, 0, 0}
	};

    program_invocation_name = program_invocation_short_name;
	setlocale (LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	while ((ch = getopt_long(argc, argv, "Vh", longopts, NULL)) != -1)
		switch (ch) {
		case 'V':
			printf(PROCPS_NG_VERSION);
			return EXIT_SUCCESS;
		case 'h':
			usage(stdout);
		default:
			usage(stderr);
		}

	argc -= optind;
	argv += optind;

	if (argc == 0)
		usage(stderr);

	pathbuf = malloc(alloclen);

	for (i = 0; i < argc; i++) {
		char *s;
		ssize_t len, buflen;
		/* Constant 10 is the length of strings "/proc/" + "/cwd" + 1 */
		char *buf;
		buflen = 10 + strlen(argv[i]) + 1;
		buf = xmalloc(buflen);

		if (check_pid_argument(argv[i]))
			xerrx(EXIT_FAILURE, _("invalid process id: %s"),
			     argv[i]);
		/*
		 * At this point, all arguments are in the form
		 * /proc/NNNN or NNNN, so a simple check based on
		 * the first char is possible
		 */
		if (argv[i][0] != '/')
			snprintf(buf, buflen, "/proc/%s/cwd", argv[i]);
		else
			snprintf(buf, buflen, "%s/cwd", argv[i]);

		/*
		 * buf contains /proc/NNNN/cwd symlink name
		 * on entry, the target of that symlink on return
		 */
		while ((len = readlink(buf, pathbuf, alloclen)) == alloclen) {
			alloclen *= 2;
			pathbuf = realloc(pathbuf, alloclen);
		}
		free(buf);

		if (len < 0) {
			s = strerror(errno == ENOENT ? ESRCH : errno);
			retval = EXIT_FAILURE;
			fprintf(stderr, "%s: %s\n", argv[i], s);
			continue;
		} else {
			pathbuf[len] = 0;
			s = pathbuf;
		}

		printf("%s: %s\n", argv[i], s);
	}
	free(pathbuf);
	return retval;
}
